src/pudl/transform/eiaaeo.py

```diff
@@ -55,6 +55,67 @@ def get_category_info(category_name: pd.Series) -> pd.Series:
     return fields
 
 
+def subtotals_match_reported_totals_ratio(
+    df: pd.DataFrame, pk: list[str], fact_columns: list[str], dimension_column: str
+) -> float:
+    """When subtotals and totals are reported in the same column, check their sums.
+
+    Group by some key, then check that within each group the non-``"total"``
+    values sum up to the corresponding ``"total"`` value.
+
+    Args:
+        df: the dataframe to investigate
+        pk: the key to group facts by
+        fact_columns: the columns containing facts you'd like to sum
+        dimension_column: the column which tells you if a fact is a sub-total
+            or a total.
+
+    Returns:
+        The ratio of reported totals that are np.isclose() to the sum of their
+        component parts.
+    """
+    reported_totals = (
+        (
+            df.loc[df[dimension_column] == "total"]
+            .set_index(pk)
+            .drop(columns=dimension_column)
+        )
+        .loc[:, fact_columns]
+        .fillna(0)
+    )
+    calculated_totals = (
+        df.loc[df[dimension_column] != "total", pk + fact_columns].groupby(pk).sum()
+    )
+    joined = calculated_totals.join(
+        reported_totals, lsuffix="_calc", rsuffix="_reported", how="inner"
+    )
+    close = np.isclose(
+        joined.loc[:, [f"{c}_calc" for c in fact_columns]],
+        joined.loc[:, [f"{c}_reported" for c in fact_columns]],
+    )
+    return close.sum() / reported_totals.size
```

Comment:
the only think i don't love about this implementation is that it will be hard to see which column in `fact_columns` is the source of the problem. and because you put the assertion outside of this function once you've hit the assertion you no longer have the detail that would enable you to figure out which fact column is the source of the problem.

Reply:
Now that you point this out, I definitely ran into this but since I was deep in the guts of the calculation already I just... looked. I'll throw a TODO in here to keep this note somewhere and maybe we'll figure out a better pattern later.